src/java/main/org/apache/zookeeper/client/HostProvider.java

```diff
@@ -53,7 +54,7 @@
      * @param spinDelay
      *            Milliseconds to wait if all hosts have been tried once.
      */
-    public InetSocketAddress next(long spinDelay);
+    public InetSocketAddress next(long spinDelay) throws UnknownHostException;
```

Comment:
That's a very good point actually, can't remember why implemented this way. 

I checked the callers' side and both fall into the same try-catch report loop which reports the problem and retry calling `next()` method. I think the reason was to keep the logic simple here and don't implement something which is already handled by someone else.

But because it's a change on a public API, it would be probably better to do the retry logic here.

Reply:
My preference is the following:

1- That we do not make this API change. It is not necessary to solve the problem of this issue.
2- That we discuss separately whether we want to change the behaviour of the `next()` in the `HostProvider` interface.

Reply:
1- See my comment below. Now I think the API change is necessary, but I need to double check how it was detected previously.
2- I started a thread on the mailing list a while ago for exactly the same reason. Let me update it.

Reply:
Some questions here @anmolnar:

>> See my comment below. Now I think the API change is necessary, but I need to double check how it was detected previously.

Which comment? I had a hard time find it, could you please explicitly paste it here?

>> I started a thread on the mailing list a while ago for exactly the same reason. Let me update it.

Similarly I lost that email thread. Could you please describe the gist of the email?

I just wan to make sure we all agree that first an API change here is needed and second, if the API changes we need throw the right type of exception. 

(Also note the current comment of HostProvider interface should be updated if the API changes, because previous semantic of next always assume returns a "valid" address.).

Reply:
@hanm Answering @fpj comment below:

> > We need a way to break the loop in the case the client closes, though.

> That's actually a good reason for _not_ dealing with the error here. Because the caller - ClientCnxn - is be able to detect client closes, but StatisHostProvider is not.

In a nutshell the problem of unresolvable DNS name must be handled somewhere and here're the considerations:
1. Client should endlessly try to resolve DNS names instead of giving up at some point. The re-try logic is already implemented in the caller methods properly.
2. In order to avoid API change we have to replicate the retry logic in the next() method which would add more complexity to the method and more extensive testing

With these considerations I strongly believe that the API change is reasonable.
That's basically what we discussed in the e-mail thread too.

I'll update the comment as requested. Thanks.

Reply:
@hanm This is done.

Reply:
>> In a nutshell the problem of unresolvable DNS name must be handled somewhere and here're the considerations

I agree with this, however this can be done without throwing the UnknownHostException here. We can just return the curAddr in case of resolving failure, and since curAddr is not resolved, trying to connect to it on caller side will not succeed and caller can continue retry etc with existing logic. The primary concern here is change of a public API which might require clients (e.g. HBase) to recompile with ZK, so I prefer the approach that can achieve same goal (let caller handle retry logic) w/o changing the signature of a public API (alternatively we can throw an unchecked exception but semantically that sounds no better than UnknownHostException).

So I am thinking the semantic of next() is:

* If we can resolve an address, resolve it and return.
* If we can't resolve an address, return an unresolved address which contain enough information (host, port etc) for caller to decide what to do.

Reply:
That's actually a great idea. I've tried that and `socket.connect()` will throw an IOException eventually of the address is not resolved. Which means basically that the same retry logic applies in the caller, it's just not next() that catches the error. I'll update the patch.

Once this one is finally merged, I'll take a look at 3.5/master branches, because I think we should be consistent with this across the board.